### DISCUSSION TITLE
countMatches syntax question

### DISCUSSION THREAD

**Discussion by AbhaMoitra** [SemanticApplicationDesignLanguage/sadl#732] [Q&A] [answered]
I am trying to use countMatches in a rule where I want to have the triple of the form "a variable of type T1". I was not able to get the syntax part correct for "type" in the following, I tried prefixing it but did not succeed.

Rule ErrorSignal2R 
if countMatches(es, "type", ErrorSignalCandidate) = 1 // does not like "type"
   and es1 is an ErrorSignalCandidate
then es1 is an ErrorSignal.

**Comment by crapo** [comment, marked as answer]
@AbhaMoitra , I'm glad you asked this question. It is a good illustration of how to use owl, rdf, or rdfs concepts that are not available in the SADL grammar in a SADL model. In some cases, the SADL grammar expresses such a concept, e.g., "Felix is a Cat" expresses the triple <Felix, rdf:type, Cat>. In your case, however, you want to express the property "rdf:type" directly as a function argument and to use it in this way you must first import an ontology that defines it. You could get the entire RDF ontology and import it, or you can create your own subset of RDF and import that. Taking the later approach, and doing the minimum, I define this simple model:
![image](https://user-images.githubusercontent.com/1186278/127552527-fa280f89-fdf6-4102-bf6a-745bc9d630f7.png). 
Note that the namespace must be correct for the RDF ontology.

Now I can import this RDF-subset ontology and use the concept "rdf:type":
![image](https://user-images.githubusercontent.com/1186278/127552761-ca7a77c7-ee4a-4195-9359-9fec2026d299.png). 
Notice the the second argument to countMatches is now known to SADL to be an RDF property as indicated by its coloring. When I run inference over this model the query returns these results:
![image](https://user-images.githubusercontent.com/1186278/127552924-996ddde4-df76-4587-9e86-7e0e61ed918f.png)

**Comment by agabaldon** [comment]
Or, since in SADL we write things like "x is a Thing" instead of "x rdf:type Thing", maybe we should have another builtin: countInstances(Thing)

**Reply by crapo** [reply]
@agabaldon , we could if there is a lot of demand for this particular property. There are many properties and classes from RDF, RDFS, and OWL that might occasionally need to be expressed in rules or as function arguments. This is not the only one nor the first time we've encountered that need.

**Reply by AbhaMoitra** [reply]
Yes, I agree that we do not want to add it as a builtin. However, maybe it could be added as a topic in http://semanticapplicationdesignlanguage.github.io/sadl/TopicalIndex.html or if we want a new "common question answer" page.

**Comment by crapo** [comment]
I've already added it to the [Topical Index](http://semanticapplicationdesignlanguage.github.io/sadl/TopicalIndex.html), under topic title [RDF, RDFS, OWL Concepts, Referencing in Expressions](http://semanticapplicationdesignlanguage.github.io/sadl/RDFConceptsInExpressions.html).
